### PR TITLE
[WIP] Documentation improvements (see #942)

### PR DESCRIPTION
**Pull request by Binero**
This PR aims to improve the documentation of the GFX crate. This is tightly bound to #942.

### CODE REVIEW

**Comment by Binero** [comment]
Fixed the typo, please don't merge yet. It's still a WIP.

File: src/render/src/encoder.rs [hunk 1]
```diff
@@ -40,8 +40,20 @@ pub enum UpdateError<T> {
...
-
-/// Graphics commands encoder.
+/// Graphics Command Encoder
+///
+/// # Overview
+/// The `Encoder` is a wrapper structure around a `CommandBuffer`. It is responsible for sending
+/// commands to the `CommandBuffer`. 
+///
+/// # Construction & Handling
+/// The `Encoder` implements `From<CommandBuffer>`, which is how it in constructed. There is no
```
Comment (kvark) [inline src/render/src/encoder.rs]: how it IS constructed

File: src/render/src/encoder.rs [hunk 2]
```diff
@@ -59,7 +71,13 @@ impl<R: Resources, C: draw::CommandBuffer<R>> From<C> for Encoder<R, C> {
...
 
 impl<R: Resources, C: draw::CommandBuffer<R>> Encoder<R, C> {
-    /// Flush the encoded commands onto the device, and clean.
+    /// Submits the commands in this `Encoder`'s internal `CommandBuffer` to the GPU, so they can
+    /// be executed. 
+    /// 
+    /// Calling `flush` before swapping buffers is critical as without it the commands of the
+    /// internal ´CommandBuffer´ will not be sent to the GPU, and as a result they will not be
+    /// processed. Calling flush too often however will result in a performance hit. It is
+    /// generally recommended to call flush once per frame, after every draw call.
```
Comment (kvark) [inline src/render/src/encoder.rs]: after every draw call?

File: src/core/src/factory.rs [hunk 3]
```diff
@@ -12,7 +12,10 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-//! Resource factory.
+//! Resource factory
+//!
+//! This module exposes the `Factory` trait, used for creating and managing graphics resources and
```
Comment (fkaa) [inline src/core/src/factory.rs]: "used for creating and managing graphics resources and" – maybe a comma before the second "and"?

File: src/core/src/factory.rs [hunk 4]
```diff
@@ -186,12 +189,37 @@ impl From<TargetViewError> for CombinedError {
...
+/// shader programs and textures. See the individual methods for more information.
+///
+/// Also see `FactoryExt` inside the `gfx` trait for additional methods.
+#[warn(missing_docs)]
 pub trait Factory<R: Resources> {
     /// Associated mapper type
     type Mapper: Clone + mapping::Raw;
 
-    /// Returns the capabilities available to the specific API implementation
+    /// Returns the capabilities of this `Factory` This usually depends on the graphics API being
```
Comment (fkaa) [inline src/core/src/factory.rs]: Missing punctuation

File: src/render/src/encoder.rs [hunk 5]
```diff
@@ -174,7 +192,7 @@ impl<R: Resources, C: draw::CommandBuffer<R>> Encoder<R, C> {
         }
     }
 
-    /// Clear a target view with a specified value.
+    /// Clears the supplied `RenderTargetView` to the supplied ´ClearColor`.
```
Comment (fkaa) [inline src/render/src/encoder.rs]: `´` at the start of ClearColor

File: src/render/src/factory.rs [hunk 6]
```diff
@@ -95,7 +102,8 @@ pub trait FactoryExt<R: Resources>: Factory<R> {
             .map_err(|e| ProgramError::Link(e))
     }
 
-    /// Create a strongly-typed Pipeline State.
+    /// Idem to `create_pipeline_from_program(..)`, but takes a `ShaderSet` as opposed to a shader
```
Comment (fkaa) [inline src/render/src/factory.rs]: "Idem" – Maybe a simpler/easier synonym?

File: src/render/src/lib.rs [hunk 7]
```diff
@@ -64,7 +64,7 @@ mod encoder;
 mod factory;
 /// Meshes
 mod mesh;
-/// Pipeline states
+// Pipeline states
```
Comment (fkaa) [inline src/render/src/lib.rs]: Missing `/`?

File: src/render/src/pso/mod.rs [hunk 8]
```diff
@@ -12,20 +12,32 @@
...
-//!    for PSO rendering, mostly consists of resource handles.
+//! # Overview
+//! A `PipelineState` holds all information needed to manage a graphics pipeline. It contains
+//! information about the shaders used, and on how to bind variables to these shaders. A 
+//! `PipelineState` manifests itself in the form of a Pipeline State Object, or PSO in short.
 //!
-//! A typed PSO is made of individual components, each is represented
-//! by the same-named field in all three generated structures.
-//! The submodules here define these components grouped by area.
+//! A Pipeline State Object exists out of different components. Every component represents
```
Comment (fkaa) [inline src/render/src/pso/mod.rs]: Add "(PSO)" after the first full mention of it

**Comment by kvark** [comment]
thanks @fkaa for having a look here!

File: src/render/src/encoder.rs (hunk 2 above)
Comment (Binero) [inline src/render/src/encoder.rs]: I meant that as in ` after (every draw call)`. I see how it is ambiguous, fixing it.

File: src/render/src/lib.rs (hunk 7 above)
Comment (Binero) [inline src/render/src/lib.rs]: This was intentional. Adding the extra / there causes it to merge with the `//!` style comments in the module itself.

**Comment by kvark** [comment]
@homu r+

**Pull request merged**